src/main/java/roomescape/reservation/service/ReservationService.java

```diff
@@ -1,79 +1,130 @@
+        ReservationSlot slot = new ReservationSlot(date, time, theme);
+
+        if (!reservationRepository.existsConfirmedReservationBySlot(slot)) {
+            throw new DataExistException("해당 시간에 예약 가능한 테마입니다. 일반 예약을 진행해주세요.");
+        }
+        if (reservationRepository.existsReservationBySlotAndMemberAndStatus(slot, member, ReservationStatus.CONFIRMED)
+                || reservationRepository.existsReservationBySlotAndMemberAndStatus(slot, member, ReservationStatus.WAITING)) {
+            throw new DataExistException("해당 시간에 이미 예약 또는 대기가 존재합나다");
+        }
+
+        return reservationRepository.save(new Reservation(member, slot, ReservationStatus.WAITING));
+    }
+
+    private void validateDate(final LocalDate date) {
         if (!date.isAfter(LocalDate.now())) {
             throw new PastDateException("과거 시간은 예약 등록을 할 수 없습니다. date = " + date);
         }
-        final ReservationTime reservationTime = reservationTimeRepository.findById(timeId)
-                .orElseThrow(() -> new DataNotFoundException("해당 예약 시간 데이터가 존재하지 않습니다. id = " + timeId));
-        final Theme theme = themeRepository.findById(themeId)
-                .orElseThrow(() -> new DataNotFoundException("해당 테마 데이터가 존재하지 않습니다. id = " + themeId));
-        if (reservationRepository.existsByDateAndTimeAndTheme(date, reservationTime, theme)) {
-            throw new DataExistException("해당 시간에 이미 예약된 테마입니다.");
+    }
+
+    @Transactional
+    public void cancelById(final Long id) {
+        Reservation reservation = reservationRepository.findById(id)
+                .orElseThrow(() -> new DataNotFoundException("해당 예약 데이터가 존재하지 않습니다. id = " + id));
+
+        if (reservation.isConfirmed()) {
+            approveNextWaitingReservation(reservation.getSlot());
         }
-        final Reservation reservation = new Reservation(member, date, reservationTime, theme);
 
-        return reservationRepository.save(reservation);
+        reservation.cancelReservation();
     }
 
-    public void deleteById(final Long id) {
-        reservationRepository.findById(id)
-                .orElseThrow(() -> new DataNotFoundException("해당 예약 데이터가 존재하지 않습니다. id = " + id));
+    @Transactional
+    public void cancelWaitingById(final Long id) {
+        Reservation reservation = reservationRepository.findById(id)
+                .orElseThrow(() -> new DataNotFoundException("해당 예약 대기 데이터가 존재하지 않습니다. id = " + id));
 
-        reservationRepository.deleteById(id);
+        reservation.cancelWaiting();
     }
 
-    public List<Reservation> findAll() {
-        return reservationRepository.findAll();
+    private void approveNextWaitingReservation(ReservationSlot slot) {
+        reservationRepository.findReservationsBySlotAndStatus(slot, ReservationStatus.WAITING)
+                .stream()
+                .findFirst()
+                .ifPresent(waitingReservation -> {
+                    waitingReservation.confirmReservation();
+                });
```

Comment:
이 케이스는 메서드 참조를 쓰면 조금 더 코드를 깔끔하게 변경할 수 있을 것 같아요. 😃

```suggestion
                .ifPresent(Reservation::confirmReservation);
```